vmaas/reposcan/database/csaf_store.py

```diff
@@ -46,7 +49,76 @@ def _save_csaf_files(self, csaf_files: CsafFiles) -> list[int]:
+                                    SELECT v.cpe, v.package, v.module
+                                    FROM (VALUES %s) AS v(cpe, package, module)
+                                    ON CONFLICT (cpe, package, module) DO NOTHING;
+                                    """, (product_tuples))
+                self.conn.commit()
+        except Exception as exc:
+            CSAF_FAILED_IMPORT.inc()
+            self.logger.exception("Failure while inserting data into csaf_products: %s", exc)
+            self.conn.rollback()
+        finally:
+            cur.close()
+
+    def _insert_cves(self, cve, product_list):
+        product_tuples = [product.to_tuple() for product in product_list]
+        try:
+            cur = self.conn.cursor()
+            if product_tuples:
+                execute_values(cur, f"""
+                    WITH product_ids AS (
+                        SELECT id
+                        FROM csaf_products
+                        WHERE (cpe, package, module) IN %s
+                    )
+                    INSERT INTO csaf_cves (cve, csaf_product_id, product_status_id)
+                    SELECT '{cve}', id, {PRODUCT_STATUS_ID}
+                    FROM product_ids
+                    ON CONFLICT (cve, csaf_product_id)
+                    DO UPDATE SET product_status_id = EXCLUDED.product_status_id;
+                    """, (product_tuples,))
+            self.conn.commit()
+        except Exception as exc:
+            CSAF_FAILED_IMPORT.inc()
+            CSAF_FAILED_UPDATE.inc()
+            self.logger.exception("Failure while inserting or updating data to csaf_cves: %s", exc)
+            self.conn.rollback()
+        finally:
+            cur.close()
+
+    def _populate_cves(self, csaf_cves):
+        for cve, product_list in csaf_cves.items():
+            self._insert_missing_products(product_list)
+            self._insert_cves(cve, product_list)
+
+    def _delete_unreferenced_products(self):
+        try:
+            cur = self.conn.cursor()
+            cur.execute("""DELETE FROM csaf_products
+                            WHERE id IN (
+                                SELECT p.id
+                                FROM csaf_products AS p
+                                LEFT JOIN csaf_cves AS c ON c.csaf_product_id = p.id
+                                WHERE c.csaf_product_id IS NULL)""")
+            self.conn.commit()
+        except Exception as exc:
+            CSAF_FAILED_DELETE.inc()
+            self.logger.exception("Failure while deleting data from csaf_products: %s", exc)
+            self.conn.rollback()
+        finally:
+            cur.close()
+
     def store(self, csaf_data: CsafData):
         """Store collection of CSAF files into DB."""
         self._save_csaf_files(csaf_data.files)
-        # TODO: populate future objects in csaf store
+        self._populate_cves(csaf_data.cves)
+        self._delete_unreferenced_products()
```

Comment:
This has a nicer API with `store` method. However, it is called with each batch (it's called about 50 times with the first/full sync). Both are fine, mentioning in case if there are still some architecture design discussions are going on.

Reply:
I'm ok with both ways, this looks fine